wasm.md

```diff
@@ -1438,20 +1440,102 @@ Start Function
+Finally, the start function is evoked.
+Exports may appear anywhere in a function, but can only be performed after what they refer to is allocated.
+Inlined imports and exports are safe to extract as they appear.
+Imports must appear first in a module due to validation.
+Exports only depend on what they are exporting having been allocated, so can appear right after allocation.
+
+`structureModule` takes a list of definitions and returns a map with different groups of definitions, preserving the order within each group.
+The groups are chosen to represent different stages of allocation and instantiation.
+
+```k
+    syntax Map ::=  structureModule (     Defns) [function]
+                 | #structureModule (Map, Defns) [function]
+                 | #initialMap ()                [function]
+ // -------------------------------------------------------
+    rule #initialMap ()
+      => "typeDecls" |-> .Defns
+         "imports"   |-> .Defns
+         "allocs"    |-> .Defns
+         "exports"   |-> .Defns
+         "inits"     |-> .Defns
+         "start"     |-> .Defns
+
+    rule structureModule(DS)
+    // The sort will build the lists in reverse, so we pre-reverse the definitons.
+      => #structureModule(#initialMap (), #reverse(DS, .Defns))
+
+    // Sorts the definitions by type, producing _reversed_ lists of the definitions.
+    rule #structureModule(M, .Defns) => M
+
+    // Type declarations.
+    rule #structureModule(M, (T:TypeDefn   DS:Defns))
+      => #structureModule(M ["typeDecls" <- (T {M ["typeDecls"]}:>Defns)], DS)
+
+    // Imports (TODO).
+ // rule #structureModule(M, (I:ImportDefn DS:Defns))
+ //   => #structureModule(M ["imports"   <- (I {M ["imports"  ]}:>Defns)], DS)
+
+    // Allocations.
+    rule #structureModule(M, (A:FuncDefn   DS:Defns))
+      => #structureModule(M ["allocs"    <- (A {M ["allocs"   ]}:>Defns)], DS)
+    rule #structureModule(M, (A:TableDefn  DS:Defns))
+      => #structureModule(M ["allocs"    <- (A {M ["allocs"   ]}:>Defns)], DS)
+    rule #structureModule(M, (A:MemoryDefn DS:Defns))
+      => #structureModule(M ["allocs"    <- (A {M ["allocs"   ]}:>Defns)], DS)
+    rule #structureModule(M, (A:GlobalDefn DS:Defns))
+      => #structureModule(M ["allocs"    <- (A {M ["allocs"   ]}:>Defns)], DS)
+
+    // Exports.
+    rule #structureModule(M, (E:ExportDefn DS:Defns))
+      => #structureModule(M ["exports"   <- (E {M ["exports"  ]}:>Defns)], DS)
+
+    // Initializations.
+    rule #structureModule(M, (I:DataDefn   DS:Defns))
+      => #structureModule(M ["inits"     <- (I {M ["inits"    ]}:>Defns)], DS)
+    rule #structureModule(M, (I:ElemDefn   DS:Defns))
+      => #structureModule(M ["inits"     <- (I {M ["inits"    ]}:>Defns)], DS)
+
+    // Start function.
+    rule #structureModule(M, (S:StartDefn  DS:Defns))
+      => #structureModule(M ["start"     <- (S .Defns)], DS) // There can only be one start function.
```

Comment:
I like this, but let's do one rule per line, so that everything lines up nicer. Also we don't need the inline comments (unless they are `TODO`). The sorts here are enough to tell us what each line does, and if there is something which needs to be there for documentation, do it in the markdown above the code block.

Reply:
Done